haskell/test/Lorentz/Contracts/Test/Permit.hs

```diff
@@ -0,0 +1,486 @@
+            callPermit stablecoinContract testPauserPK testPauserSK 0 Pause
+
+          withSender wallet2 $
+            lCallEP stablecoinContract (Call @"Pause") ()
+
+    specify "Permits are consumed after use" $
+      integrationalTestExpectation $ do
+        withOriginated originate defaultOriginationParams $ \stablecoinContract -> do
+          withSender testPauser $ do
+            callPermit stablecoinContract testPauserPK testPauserSK 0 Pause
+            callPermit stablecoinContract testPauserPK testPauserSK 1 Unpause
+
+          withSender wallet1 $ do
+            lCallEP stablecoinContract (Call @"Pause") ()
+            lCallEP stablecoinContract (Call @"Unpause") ()
+            lCallEP stablecoinContract (Call @"Pause") () `catchExpectedError` mgmNotPauser
+
+    specify "Admins do not consume permits" $
+      integrationalTestExpectation $ do
+        withOriginated originate defaultOriginationParams $ \stablecoinContract -> do
+
+          withSender testPauser $ do
+            callPermit stablecoinContract testPauserPK testPauserSK 0 Pause
+            lCallEP stablecoinContract (Call @"Pause") ()
+            lCallEP stablecoinContract (Call @"Unpause") ()
+          withSender wallet1 $
+            lCallEP stablecoinContract (Call @"Pause") ()
+
+    specify "Counter is increased every time a permit is issued" $ do
+      integrationalTestExpectation $ do
+        withOriginated originate defaultOriginationParams $ \stablecoinContract -> do
+          let issuePermit counter =
+                void $ callPermit stablecoinContract testPauserPK testPauserSK counter Pause
+
+          withSender testPauser $ do
+            issuePermit 0
+            issuePermit 1
+            issuePermit 2
+            issuePermit 2 `catchExpectedError` errMissignedPermit
+            issuePermit 3
+
+    specify "`Unpause` permit cannot be used to pause" $
+      -- More generally, we want to assert that if two entrypoints X and Y have the same
+      -- parameter type (in this case, both `Pause` and `Unpause` are of type `Unit`),
+      -- then a permit issued for X cannot be used to access entrypoint Y.
+      integrationalTestExpectation $ do
+        withOriginated originate defaultOriginationParams $ \stablecoinContract -> do
+          withSender wallet2 $ do
+            callPermit stablecoinContract testPauserPK testPauserSK 0 Unpause
+            lCallEP stablecoinContract (Call @"Pause") () `catchExpectedError` mgmNotPauser
+
+    specify "Permits expire after some time (set by the contract)" $ do
+      integrationalTestExpectation $ do
+        withOriginated originate defaultOriginationParams $ \stablecoinContract -> do
+          let defaultExpiry = opDefaultExpiry defaultOriginationParams
+
+          withSender testPauser $ do
+            callPermit stablecoinContract testPauserPK testPauserSK 0 Pause
+          withSender wallet1 $ do
+            rewindTime (fromIntegral defaultExpiry + 1)
```

Comment:
@gromakovsky Seems like we've found a good use case for advancing time in integration tests ^^

I'll un-deprecate this in `cleveland`, and add an `advanceTime :: Natural -> ...` function to Nettest (as part of `#53`).

Reply:
> un-deprecate this

Only `rewindTime` (and rename to `advanceTime` and change its argument to be unsigned) or other time-related functions as well? I don't remember the whole story, but I think we decided that we can't reach feature parity between Nettest and Integrational, so there will be some functions only present in Integrational and it will be possible to write a test that only works with integrational. Is it right? If it's right, why did we decide to deprecate `setNow` then? Possible reasons:
1. When we decided to deprecate it, we still were hoping it would be possible to reach feature parity.
2. We decided that time manipulations other than advancing time are rarely needed, hence it's better to force users to use functions that can be implemented in both Integrational and Client nettests.

Also note that timestamps returned by `NOW` are quite descrete. AFAIU, it always returns timestamp of a block (or `f(timestamp of a block)`. So for example waiting for 10 seconds doesn't make much sense: it can be too much if blocks are created every second or can have no effect if blocks are created once a minute. So this `Natural` argument should be a number of blocks, I guess, not a time unit (like seconds). Maybe that's what you had in mind anyway.

Reply:
> why did we decide to deprecate setNow then

It was (2).
Back then, we decided that:

* Time-related functions were only ever used in some `nbit` tests, which could be re-written to avoid these functions. So we thought the easiest solution was to deprecate these functions, re-write the tests to avoid them and to use `Nettest` instead of `Integrational`.
* `Branchout`/`offshoot` would be added to `Nettest` in such a way that they could only be used in an emulated env.

Seeing as we've found a scenario where advancing time is _required_ (in order to test permit expiries), I propose we do for `advanceTime` what we decided to do for `branchout`/`offshoot`. (We could also close NBIT-44).

> So this Natural argument should be a number of blocks, I guess, not a time unit (like seconds).

I'm not sure what you meant here. The use case here is we want to test that a permit is no longer usable after `n` seconds have elapsed. I don't see how advancing `n` blocks would achieve this. Also, I don't see how this would be implemented in morley, since `Michelson.Runtime`\`Michelson.Interpret` have no concept of blocks.

Reply:
>  I propose we do for `advanceTime` what we decided to do for `branchout`/`offshoot`.

Ah, ok, I didn't realize that. So you propose to implement `advanceTime` only in Pure, not in Client. I think there is nothing preventing us from doing `advanceTime` in Client as well. We can just `threadDelay` there. The only problem is that in Client time is discrete. If you make two contract calls and want time returned by `NOW` to differ by 25 seconds (between these two calls), you can't just do `threadDelay (25 sec)` because:
1. In carthagenet blocks are created once per 30 seconds, in mainnet – once per 1 minute. So the difference can be more than 25 seconds.
2. We can say that `advanceTime x` advances time by _at least_ `x`, but we still should be careful because timestamp is taken from a block created by an arbitrary node and that node's clock can be different. I guess there is some verification of timestamps, but it's quite approximate. So if at 18:00:00 someone calls `advanceTime (1 min)` and we wait for 1 minute, we may receive block that has `18:00:59` in it (e. g. at 18:01:01) which would be incorrect. So we have to wait for a block and then maybe wait again.

So there is some complexity here, but I think it's totally possible to have `advanceTime` in both Client and Pure. And it would be nice to have it in Client because it is something that can work fine in emulated environment, but work incorrectly in real network.

There can be cases when we don't want to run tests with `advanceTime` in Client implementation. Most importantly: when it is called with a large argument, e. g. a minute or more. I think in most cases there is no need to wait for a long value, all periods can be scaled. Maybe ideally we should make it possible to say "this test will take too much time in Client, please only use Pure for it", but I guess it's not very important.

Regarding `Natural` argument – now I think it should be a time unit, and the fact that time is discrete should be handled differently (e. g. by saying that `advanceTime x` waits for `≥ x`).